### PR TITLE
feat(pid_longitudinal_controller): improve delay and slope compensation

### PR DESCRIPTION
**Pull request by brkay54**
## Description

In the current implementation on Autoware, while calculating the output command, we are compensating for the slope and delay in `pid_longitudinal_controller`. 

We are doing the delay compensation by using the estimation of the drive distance of the vehicle after `delay_compensation_time` and we are looking at the future `TrajectoryPoint` and calculating the output command with respect to this future point. 

However, before sending the output command, we are also doing slope compensation with respect to **Current** `TrajectoryPoint`. However, in some scenarios on hilly roads, this way is causing some conflicts because of the time like below.

delayslope drawio

As you can see in this scenario, there is a conflict between slope and delay compensation because they are using different reference times. In this scenario, the node will send FF with respect to the target point however it sums negative slope compensation value because of the current position of the vehicle and it causes undesired behavior. To avoid this behavior, I made some changes to delay and slope compensation in this package to sum compensations at the same time.

Also, after this PR, the following PRs should be merged by order here:

1- https://github.com/autowarefoundation/autoware.universe/pull/5077
2- https://github.com/autowarefoundation/autoware.universe/pull/5079
3- https://github.com/autowarefoundation/autoware.universe/pull/5080
4- https://github.com/autowarefoundation/autoware.universe/pull/5081

## Tests performed

I tested on the real vehicle and also on the planning simulator. It solves our problems in the real environment and there is no behavioral change in the planning simulator.

## Effects on system behavior

The only effect is that we will use the raw pitch info at only starting (while not moving), if the vehicle is moving, it needs the future pitch info, to achieve this, it depends on the pitch values of the trajectory points.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### Summary by CodeRabbit

- New Feature: Improved delay and slope compensation in the `pid_longitudinal_controller` package, resulting in more accurate output commands. The changes include updates to the `getPitchByTraj` function, introduction of the `findTrajectoryPoseAfterDistance` function, and modification of the return type of the `lerpTrajectoryPoint` function. These enhancements may impact the behavior of the code.
- Bug Fix: Addressed delay and slope compensation issues in the `PidLongitudinalController` class of Autoware. The modifications involve using drive distance estimation after a specified delay time for delay compensation and adjusting slope compensation to avoid conflicts. The changes also introduce the `StateAfterDelay` struct and a new `SlopeSource` enum.

### CODE REVIEW

**Comment by TakaHoribe** [comment]
@brkay54 Thank you for this PR. Recent PRs have caused conflicts. Would you fix it? Sorry for your inconvenience.

**Comment by brkay54** [comment]
Hi @TakaHoribe sorry for the conflicts. I updated all the PRs.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:434]: Do we need casting to `uint`? We can use `size_t` directly, I guess.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:433]: When you add a new point in the trajectory, please ensure there are no duplicated points. Otherwise, some computation throws an exception.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:459]:
```suggestion
  // calculate the target motion for delay compensation
```

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:716]: Why this is moved? I think this function is necessary only for DRIVE state.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:821]: I don't know why the slope compensation causes a sudden stop. The sudden acceleration should not depend on the slope compensation, but the stop acceleration parameter, I guess.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:939]:
```suggestion
    running_distance = std::abs(
```

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:962]:
```suggestion
        std::abs(std::abs(pred_vel) * time_to_next_acc + 0.5 * acc * time_to_next_acc * time_to_next_acc);
```

**Review by TakaHoribe — COMMENTED** [review]
@brkay54 Thank you for this PR. The overall strategy looks god.
Would you check my comments? Thank you!

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:496]: Regarding which slope information to use, it should not depend on the logic as a basic policy because it depends on the accuracy of sensors and maps. What is the purpose of this implementation?

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:947]: Is it ok to use `m_delay_compensation_time`? Should it be `delay_compensation_time`? There are same implementations in this function.

**Review by TakaHoribe — COMMENTED** [review]
other comments.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (brkay54) [reply control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:166]: Because slope compensation won't be done with respect to the user's option. While driving the ego, it is always needed that the slope should be calculated from future position projection to the trajectory of the vehicle.

The pitch information from the sensing only is usable when the vehicle is in the stopped state.

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (brkay54) [reply control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:496]: > Regarding which slope information to use, it should not depend on the logic as a basic policy because it depends on the accuracy of sensors and maps. What is the purpose of this implementation?



Actually, the pitch information from sensing would be more reliable from the trajectory. My purpose is, that if the vehicle is stopping, it can use the pitch information from sensing, with no need for future slope. But if the vehicle is moving (running distance after delay >0.0), it needs pitch information from the future.

While changing the pitch input source, to provide smoother change, I put them into LPF.

Maybe the slope input source should change with respect to running_distance threshold like:

if (running_distance > 0.5?) {
control_data.slope_angle = traj_pitch;
m_lpf_pitch->filter(control_data.slope_angle);
} else {
control_data.slope_angle = raw_pitch;
// send filtered slope angle to smooth transition between raw and traj pitch
control_data.slope_angle = m_lpf_pitch->filter(control_data.slope_angle);
}

Because in the current implementation, it is possible that the vehicle's velocity is < 0.001 but the vehicle may be in the DRIVE state. What do you think?

**Comment by brkay54** [comment]
Hi @TakaHoribe thank you for your review. I updated the branch w.r.t. your feedback. 
Just I didn't update the code snippets related to:
https://github.com/autowarefoundation/autoware.universe/pull/4712#discussion_r1342496103
https://github.com/autowarefoundation/autoware.universe/pull/4712#discussion_r1342528866 
I am waiting for your feedback! Thank you so much for your effort ❤️

**Comment by brkay54** [comment]
Hi @TakaHoribe, sorry to disturb you, but just I want to kindly ping you.

File: control/pid_longitudinal_controller/src/longitudinal_controller_utils.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/longitudinal_controller_utils.cpp:198]: This would be more simple?

```suggestion
      const auto p0 = trajectory.points.at(i).pose;
      const auto p1 = trajectory.points.at(i).pose;
      p.position.x = interpolation::lerp(p0.position.x, p1.position.x, ratio);
      p.position.y = interpolation::lerp(p0.position.y, p1.position.y, ratio);
      p.position.z = interpolation::lerp(p0.position.z, p1.position.z, ratio);
      p.orientation = lerpOrientation(p0.orientation, p1.orientation, ratio);
```

**Comment by TakaHoribe** [comment]
@brkay54 Sorry for my late reply 

> Actually, the pitch information from sensing would be more reliable from the trajectory. 

This depends on the users. For the pitch source, we still need to keep the sensor-only option for users who can not have an accurate map.

So I would suggest this way:

Provide three options: 

- `vehicle_pitch`: Use the vehicle's pitch value as a slope. The information derived from the vehicle's pitch is often not an accurate representation of the road gradient due to effects such as body vibrations.
- `trajectory` : Use the trajectory slope value as a slope. Calculating the precise road gradient is possible with trajectory data; however, it requires high-quality map information.
- `trajectory_adaptive`: Primarily, the gradient is calculated using trajectory data, but if the gradient cannot be computed at the end of the route or similar situations, the vehicle pitch is used instead.

The default could be `vehicle_pitch` or `trajectory_adaptive`.

What do you think?

**Comment by brkay54** [comment]
@TakaHoribe Thank you for the suggestion. It makes sense to me. I think it is better to make `trajectory_adaptive` as the default option because it can work with delay compensation better. 

For the trajectory_adaptive option, if it is okay, I am planning to make it adaptive with respect to the velocity of the vehicle. It will use the vehicle_pitch in low velocities (because in low velocities, the target trajectory point gonna be very close to the vehicle).

**Comment by brkay54** [comment]
> @TakaHoribe Thank you for the suggestion. It makes sense to me. I think it is better to make `trajectory_adaptive` as the default option because it can work with delay compensation better.
> 
> For the trajectory_adaptive option, if it is okay, I am planning to make it adaptive concerning the velocity of the vehicle. It will use the vehicle_pitch in low velocities (because in low velocities, the target trajectory point gonna be very close to the vehicle).

Hi again @TakaHoribe -san, I wrote like this but after thinking more, IMO, 2 options would be better rather than using 3 options.

Using only trajectory has a problem for the last point, as you said, and also for all cases, we will have vehicle pitch value. I think the `trajectory` option is not necessary because we will need the future slope only for higher velocities, for lower velocities we can use vehicle_pitch (BCS target waypoint will be near ego). If we do not assume that the user may not have vehicle pitch value, the `adaptive_trajectory` option is always the better option than `trajectory`. 

My proposal is that like that:

```
# slope compensation
adaptive_trajectory: true  # if true, trajectory is used as pitch information source for higher velocities and vehicle pitch for lower velocities. if false, vehicle_pitch is used as pitch information.
adaptive_trajectory_treshold: 1.0 # available only adaptive_trajectory is true
```
What do you think?

**Comment by brkay54** [comment]
Hi again @TakaHoribe -san, I updated the PR w.r.t my comment above. Also, I updated the launch [PR](https://github.com/autowarefoundation/autoware_launch/pull/585) and readme [PR](https://github.com/autowarefoundation/autoware.universe/pull/5081). Please let me know if you have any concerns.

**Comment by TakaHoribe** [comment]
@brkay54 Thank you for addressing my comments. 

> because we will need the future slope only for higher velocities, for lower velocities we can use vehicle_pitch

This is only in the ideal case. We need to care about users who can not use an accurate map or rich sensor for pitch estimation. 

Trajectory-only option is necessary for users with poor sensors. 
Pitch-only option is necessary for uses with poor maps.
Adaptive can be available only for users with both ich sensors and rich maps.

**Comment by brkay54** [comment]
Hi @TakaHoribe -san, sorry to be late I couldn't find time to update the PR in the last weeks. 

I updated the slope source option as you said above. You can find the changes in the launch file [here](https://github.com/autowarefoundation/autoware_launch/pull/585). It is ready for review, please let me know if any issue exists.

Also, after this PR, the following PRs should be merged by order here:

1- https://github.com/autowarefoundation/autoware.universe/pull/5077
2- https://github.com/autowarefoundation/autoware.universe/pull/5079
3- https://github.com/autowarefoundation/autoware.universe/pull/5080
4- https://github.com/autowarefoundation/autoware.universe/pull/5081

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (TakaHoribe) [inline control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp:452]: `nearest_idx` is not always equivalent to `segment_idx + 1`. 
But the `segment_idx + 1` will be `nearest_idx` when the point is inserted in the `interpolated_traj`.

I guess this variable is not used anywhere?

**Review by TakaHoribe — COMMENTED** [review]
@brkay54 LGTM and almost ready to merge.

Would you address the followings?

- Fix test code. This seems to fail due to the pos/neg definition mismatch. 
- Make a PR to the autoware_launch for new parameters.
- Make sure unrelated parameters are not changed in this PR: parameters attached in the screenshot should not change.

image

File: control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp
Comment (brkay54) [reply control/pid_longitudinal_controller/src/pid_longitudinal_controller.cpp]: Thank you for the review Horibe-san. Because it will be inserted into interpolated_traj, I set current_interpolated_pose.second + 1` to `nearest_idx` as below:

```
  // Insert the interpolated point
  control_data.interpolated_traj.points.insert(
    control_data.interpolated_traj.points.begin() + nearest_idx, nearest_point);
  control_data.nearest_idx = nearest_idx;
  control_data.target_idx = control_data.nearest_idx; // target_idx will be calculated, we set nearest_idx for now.
  ```
  
  Yes, the variable is not used anywhere, I simplified it. 
  
> `nearest_idx` is not always equivalent to `segment_idx + 1`.

I am not sure I can understand what you mean. I defined it like that because it will be inserted into interpolated_traj (If I understand correctly). I simplified this snipped. Please let me know if there is any problem.

>     * Fix test code. This seems to fail due to the pos/neg definition mismatch.

The test file is updated.

>     * Make a PR to the autoware_launch for new parameters.

I created PR for the parameter update: https://github.com/autowarefoundation/autoware_launch/pull/585
 
>     * Make sure unrelated parameters are not changed in this PR: parameters attached in the screenshot should not change.

Sorry for the changes, I updated.

**Comment by brkay54** [comment]
@TakaHoribe-san, Just to remind: Also, after this PR, the following PRs should be merged by order here:

1- https://github.com/autowarefoundation/autoware.universe/pull/5077
2- https://github.com/autowarefoundation/autoware.universe/pull/5079
3- https://github.com/autowarefoundation/autoware.universe/pull/5080
4- https://github.com/autowarefoundation/autoware.universe/pull/5081

Also, Horibe-san, I made some critical changes to the package. If you don't mind, I would like to maintain this package. Can I add myself as a maintainer?

**Review by TakaHoribe — APPROVED** [review]
LGTM!

**Comment by TakaHoribe** [comment]
@brkay54 

> Can I add myself as a maintainer?

I think it is fine. I'll talk with @takayuki5168

**Pull request merged**
